grype/version/constraint_expression.go

```diff
@@ -61,3 +71,49 @@ func (c *constraintExpression) satisfied(other *Version) (bool, error) {
 	}
 	return oneSatisfied, nil
 }
+
+func scanExpression(phrase string) ([][]string, error) {
+	var scnr scanner.Scanner
+	var orGroups [][]string // all versions a group of and'd groups or'd together
+	var andGroup []string   // most current group of and'd versions
+	var buf bytes.Buffer    // most current single version value
+	var lastToken string
+
+	captureVersionOperatorPair := func() {
+		if buf.Len() > 0 {
+			ver := buf.String()
+			andGroup = append(andGroup, ver)
+			buf.Reset()
+		}
+	}
+
+	captureAndGroup := func() {
+		if len(andGroup) > 0 {
+			orGroups = append(orGroups, andGroup)
+			andGroup = nil
+		}
+	}
+
+	scnr.Init(strings.NewReader(phrase))
+	tokenRune := scnr.Scan()
+	for tokenRune != scanner.EOF {
+		currentToken := scnr.TokenText()
+		switch {
+		case currentToken == ",":
+			captureVersionOperatorPair()
+		case currentToken == "|" && lastToken == "|":
+			captureVersionOperatorPair()
+			captureAndGroup()
+		case currentToken == "(" || currentToken == ")":
+			return nil, fmt.Errorf("parenthetical expressions are not supported yet")
+		case currentToken != "|":
+			buf.Write([]byte(currentToken))
+		}
+		lastToken = currentToken
+		tokenRune = scnr.Scan()
+	}
+	captureVersionOperatorPair()
```

Comment:
I haven't seen this anonymous function usage before, and I'm having trouble following what is doing. Is it any different from adding the `if` conditions from the anonymous functions here?

Reply:
indeed, no different that what you mentioned... it was to save on code repetition was all. For these two lines in particular it is possible for the loop to have been processing a token (since there will not necessarily be trailing `,` or `||` tokens) so this is to capture whatever is left in the buffer and the last `and` group into the final results.

Reply:
Ahhh of course! that makes sense. Thanks for clarifying this, I'd not seen this pattern before in Go.